hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java

```diff
@@ -124,6 +125,28 @@ public static void validateAndAddRequiredKeys(
         }
     }
 
+    /**
+     * The function is unique for File delete Pre-handle validates the keys.
+     * Then it wraps the file Keylist into thresholds adds them to the context.
+     *
+     * @param file file that will be checked for required keys
+     * @param context the prehandle context for the transaction.
+     */
+    public static void validateAndAddRequiredKeysForDelete(
+            @Nullable final File file, @NonNull final PreHandleContext context) {
+        if (file != null) {
+            KeyList fileKeyList = file.keys();
+
+            if (fileKeyList != null && fileKeyList.hasKeys()) {
+                // this threshold created to solve the issue of the file delete that can be deleted using one of the
+                // keys in the keylist that's why it is wrapped in threshold key with threshold 1
+                ThresholdKey syntheticKey =
+                        ThresholdKey.newBuilder().threshold(1).keys(fileKeyList).build();
+                context.requireKey(Key.newBuilder().thresholdKey(syntheticKey).build());
+            }
+        }
```

Comment:
I think there is a bug here. If File is null, or if it has no keys, then we don't add a "requiredKey". But then this means it is by default accepted. So if the File is null for some reason, the transaction will succeed? Or if the file has no keys on it, instead of forbidding modification, it will be allowed. Is that right?

Reply:
in File handle we check `validateFalse(file.keys() == null, UNAUTHORIZED);`
The function you commented about is just for prehandle when we add required keys.

Reply:
It seems like we could just throw a `PreCheckException(INVALID_FILE_ID)` on a missing `File`, as e.g. `TokenDeleteHandler` does [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenDeleteHandler.java#L60).

Reply:
This isn't checking for missing file, it's handling that delete only requires a single key to succeed.  Missing file is checked elsewhere, and already throws as described.

Reply:
That exactly what is happens in the function that was called before that
`preValidate(transactionFileId, fileStore, context, true);`

``` public static void preValidate(
            @Nullable final FileID fileId,
            @NonNull final ReadableFileStore fileStore,
            @NonNull final PreHandleContext context,
            boolean isDelete)
            throws PreCheckException {
        requireNonNull(context);

        if (fileId == null) {
            throw new PreCheckException(INVALID_FILE_ID);
        }

        final var fileConfig = context.configuration().getConfigData(FilesConfig.class);```

Reply:
@jsync-swirlds 
Regarding your concerns, the list of required keys we get in pre-handle only triggers the signature verification asynchronously. We do not use this information for anything else. In handle, we get the required keys again (this time from the right state), and based on that information, we verify all required keys are valid.

Reply:
Ahh, that's the second call to preHandle in the Handle flow.  I forgot about that.
Thank you, @netopyr .